velox/exec/LocalPlanner.cpp

```diff
@@ -487,4 +535,50 @@ std::shared_ptr<Driver> DriverFactory::createDriver(
 
   return std::make_shared<Driver>(std::move(ctx), std::move(operators));
 }
+
+std::vector<core::PlanNodeId> DriverFactory::needsHashJoinBridges() const {
+  std::vector<core::PlanNodeId> planNodeIds;
+  // Ungrouped execution pipelines need to take care of cross-mode bridges.
+  if (!groupedExecution && !mixedExecutionModeHashJoinNodeIds.empty()) {
+    planNodeIds.insert(
+        planNodeIds.end(),
+        mixedExecutionModeHashJoinNodeIds.begin(),
+        mixedExecutionModeHashJoinNodeIds.end());
+  }
+  for (const auto& planNode : planNodes) {
+    if (auto joinNode =
+            std::dynamic_pointer_cast<const core::HashJoinNode>(planNode)) {
+      // Grouped execution pipelines should not create cross-mode bridges.
+      if (!groupedExecution ||
```

Comment:
Hi @spershin , seems we should use && instead of || here? Or else it may insert duplicate planNode into planNodeIds.

Reply:
@waitinfuture 
Thanks for paying attention to this.
This code is quite confusing. Took me 5 mins to recall and figure out. :)

I believe the code is correct as is.
Will try to explain.
Here we play with two bool variables:
`isGroupedExecutionPipeline` and `doesNodeCorrespondToBridgeConnectingGroupedAndUngroupedPipelines`
So, we have 4 combinations for each node we look at when generating return result of needsHashJoinBridges():
pipeline_grouped  node_is_in_mixed_list
true                         true
true                         false
false                        true
false                        false

We know that ungrouped pipelines should create 'mixed' bridges and grouped pipelines should not.

So
pipeline_grouped  node_is_in_mixed_list   create_bridge
true                         true                                  NO, because bridge is mixed and pipeline is grouped.
true                         false                                 yes, because bridge is not mixed (grouped) and pipeline is grouped.
false                        true                                  yes, because bridge is mixed and pipeline is ungrouped.
false                        false                                 yes, because bridge is not mixed (ungrouped) and pipeline is ungrouped.

So, the "if" condition before the "for" loop takes care of case 3 for ungrouped pipeline. Correct.
The condition in the loop takes care of the rest: 1, 2 and 4.
For 1 it will not add the node to the vector because the node is in the mixed list. Correct
For 2 it will add the node to the vector because the node is not in the mixed list. Correct.
For 3 it will add any node to the vector, but none of this nodes will be in the mixed list, because any node in the mixed list comes only from the grouped pipeline - that's why we have that list in the 1st place. I assume you meant this duplication. Correct.
For 4 it will add the node to the vector because the pipeline is ungrouped and node is not mixed anyways. Correct.

Let me know if this is not clear or I'm still not seeing something.
Creating a unit test (based on the existing ones for simplicity) can also prove or disprove it.

Reply:
Hi @spershin , thanks for your detailed explanation! After rethinking for several times, I finally understand. Key is ```
but none of this nodes will be in the mixed list, because any node in the mixed list comes only from the grouped pipeline```

I missed the point that for any node in a ungrouped pipeline's mixed list, it is from the grouped pipeline, not from itself.

Thanks!